src/mgr/InfoAction.cc

```diff
@@ -143,14 +143,17 @@ Mgr::InfoAction::dump(StoreEntry* entry)
     Must(entry != nullptr);
 
 #if XMALLOC_STATISTICS
```

Comment:
> [Francesco](https://github.com/squid-cache/squid/pull/1784/files#r1587671176): InfoAction is a special case (and not in a good way): it has a section which is not aggregated, and a section which is.

This is an important discovery because the current [proposal](https://github.com/squid-cache/squid/pull/1784#issuecomment-2088547341) may not be compatible with such "combination" reports. Before we decide how to deal with it, let me ask you this question: Does the current _official_ `'mgr:info` report work correctly when XMALLOC_STATISTICS is enabled? Could you share the corresponding SMP and non-SMP report samples from the official Squid, please?

Reply:
We need to define what "working" means. The output it produces is arguably correct, but the format it has is:
```
---
kid: 1

Memory allocation statistics
  Alloc Size           Count  Delta    Alloc/sec
           4             688    688 0.0
           8             422    422 0.0
          12             372    372 0.0
[...]
      522864               1      1 0.0
...
---
kid: 2

Memory allocation statistics
  Alloc Size           Count  Delta    Alloc/sec
           4             688    688 0.0
           8             422    422 0.0
[...]
...
[ repeat for all workers]
...
---
Squid Object Cache: Version 7.0.0-VCS
Build Info:
Service Name: squid
Start Time:     Fri, 03 May 2024 08:20:39 GMT
Current Time:   Fri, 03 May 2024 08:40:31 GMT
Connection information for squid:
        Number of clients accessing cache:      2
        Number of HTTP requests received:       6
        Number of ICP messages received:        0
        Number of ICP messages sent:    0
        Number of queued ICP replies:   0
        Number of HTCP messages received:       0
        Number of HTCP messages sent:   0
        Request failure ratio:   0.00
        Average HTTP requests per minute since start:   0.3
        Average ICP messages per minute since start:    0.0
        Select loop called: 8301 times, 430.705 ms avg
Cache information for squid:
        Hits as % of all requests:      5min: 0.0%, 60min: 0.0%
        Hits as % of bytes sent:        5min: 0.0%, 60min: 0.0%
        Memory hits as % of hit requests:       5min: 0.0%, 60min: 0.0%
        Disk hits as % of hit requests: 5min: 0.0%, 60min: 0.0%
        Storage Swap size:      16 KB
        Storage Swap capacity:   0.0% used, 100.0% free
        Storage Mem size:       0 KB
        Storage Mem capacity:    0.0% used, 100.0% free
        Mean Object Size:       0.00 KB
        Requests given to unlinkd:      0
Median Service Times (seconds)  5 min    60 min:
        HTTP Requests (All):   0.00000  0.00000
        Cache Misses:          0.00000  0.00000
        Cache Hits:            0.00000  0.00000
        Near Hits:             0.00000  0.00000
        Not-Modified Replies:  0.00000  0.00000
        DNS Lookups:           0.00000  0.00000
        ICP Queries:           0.00000  0.00000
Resource usage for squid:
        UP Time:        1191.752 seconds
        CPU Time:       0.448 seconds
        CPU Usage:      0.04%
        CPU Usage, 5 minute avg:        0.02%
        CPU Usage, 60 minute avg:       0.04%
        Maximum Resident Size: 308336 KB
        Page faults with physical i/o: 24
[etc]
```

In other words, it doesn't follow the framing conventions we are converging to, and it doesn't really make logical sense: "info" associates in mu mind with a high level overview of the service state, not low-level details about how memory allocation works (granted, malloc stats are compile-time optional).

What I think would make sense is to take malloc statistics into their own optional report, thus also cleaning up the report logical structure

Reply:
> it doesn't follow the framing conventions we are converging to, and it doesn't really make logical sense: "info" associates in mu mind with a high level overview of the service state, not low-level details about how memory allocation works

Agreed on both counts.

> What I think would make sense is to take malloc statistics into their own optional report, thus also cleaning up the report logical structure

Probably, but let's ignore mgr:info specifics for a second and answer a much bigger/general question: Do we want Squid (to continue) to support "mixed" reports (i.e. reports that contain a mixture of non-aggregated and aggregated information)? 

Pros: Mixed reports allow us to naturally express related Squid statistics/state without artificially splitting a report into two (aggregated plus kid-specific).

Pros: Allowing mixed reports goes inline with allowing different kids to report different kinds of stats/state in per-kid reports, which is already happening in some reports for kids playing roles (e.g., worker vs disker). We would be essentially extending that model to its logical conclusion: From "a report is a stream of documents with related but possibly different per-kid info type" to "a report is a stream of documents with related by different [aggregation and/or per-kid] info type".

Cons: Squid reporting code stays (or will become) more complex compared to the simpler "mixing is banned" model.

Cons: The consumer should prepare to receive aggregated info mixed with kid-specific reports in virtually _any_ cache manager report. This requires more consumer work/code in some cases (e.g., simply counting documents in a stream may not yield the number of reporting kids).

If the answer is "allow mixed reports", then we need to revise the stream-formatting/document-identification proposal to accommodate such reports.

Reply:
I'd say "no mixed reports" in this context.
As in: I don't think we should increase the infrastructure complexity by allowing mixed reports. If a specific report('s developer) wants to display per-kid statistics in their report because it makes sense in that specific report's context, they're free to do so but it is that report's responsibility

Reply:
> I don't think we should increase the infrastructure complexity by allowing mixed reports. If a specific report('s developer) wants to display per-kid statistics in their report because it makes sense in that specific report's context, they're free to do so but it is that report's responsibility. From Squid's perspective it is a regular aggregate report

N.B. I will try to use "kid-reported" instead of "per-kid" terminology in this context. It is the actor that doing the reporting that matters here the most when talking about infrastructure and code design, not the fact that the reported information happens to be specific to some kid. It is the shoveling of client connection descriptor and associated cache manager transaction state from the receiving kid to Coordinator and from Coordinator to the reporting kid that creates a lot of code complexities and design limitations...

Since any reasonable display of kid-reported statistics requires infrastructure/code to get the kid to report that statistics, I disagree that we can punt on this issue by leaving it to report developer's discretion: Our general infrastructure/code either supports mixed reports (well) or it does not. For example, if mixed reports are supported (when developers want to produce them) then the current (complex) Mgr::Action::aggregatable() description (quoted below) is accurate. Otherwise, that description (and probably current or future the code using aggregatable()!) can be simplified.

```C++
/// whether at least some local kid info can be combined and, hence, the
/// combined data should be written at the end of the coordinated response
virtual bool aggregatable() const { return true; } // most kid classes are
```

> From Squid's perspective it is a regular aggregate report

A similar concern applies to the mixed report _consumer_. If developers are allowed to create mixed reports, then

* If we let the report developer insert YAML documents (one for each reporting kid) then the mixed report will not match the formatting/tagging we are documenting. This mismatch is clearly bad for consuming applications.
* Otherwise (i.e., if we do not let the report developer insert kid-reported YAML documents), then the developer would be forced to write complicated code that allows reporting kid to artificially over-indent its YAML, so that it can be integrated into the surrounding report document. This kind of hacks is exactly what we are trying to prevent by using YAML documents for kid-reported content in the first place!

Either way, we end up with a bad outcome.

To me, it feels like we need to relax formatting/tagging specs to allow (certain) mixed reports. I do not think it will be very difficult to do that. If you agree, I can take a stab at it.

Reply:
> If we let the report developer insert YAML documents (one for each reporting kid) then the mixed report will not match the formatting/tagging we are documenting. This mismatch is clearly bad for consuming applications.

This is the point I'm trying to make. A report is either a series of documents, each produced by a kid, or it is a single document, produced by aggregating data from the kids. This aggregated document needs to be yaml, but nothing prevents it from having a section which - say - is structured like:
```
some value:
  kid1: number1
  kid2: number2
  etc 
```
We cannot foresee what these reports could look like, because right now we have none. This is why I'd leave this as something to be addressed when a good use case will come out

This said, I'm not going to ask you to not do anything :) As long as the project keeps on moving, I'm happy :)

Reply:
> Francesco: A report is either a series of documents, each produced by a kid, or it is a single document, produced by aggregating data from the kids.

The above matches our draft specs but does not match existing Squid code. Existing Squid code is designed and written to support what I called, for the lack of a better term, "mixed reports": a mixture of documents where some documents are produced by kids and some documents are produced by the aggregator.

> This aggregated document needs to be yaml, but nothing prevents it from having an [aggregator-produced] section [with kid-specific info].

You are correct, but that particular benign use case is _not_ the problematic use case we need to address. The use case we need to address is different. That different use case cannot be addressed by the same (correct) "document producer decides what information to include" principle that applies to the benign use case. The use case we need to address is, for example, a report containing three documents: one document produced by kid 1, one document produced by kid 2, and a third document produced by the aggregator:

```
---
kid: 1
<some info produced by kid 1>
...
---
kid: 2
<some info produced by kid 2>
...
---
kids: [1, 2]
<some info produced by the aggregator>
```

* Squid code is designed to support such mixed reports today.
* There is at least one existing report that has that kind of mixed structure. 
* Yet, our daft specs (and the proposed rule quoted at the top of this message) ban such reports.

Thus, we need to decide whether we start restricting/simplifying Squid code (to match the draft specs) OR relax our draft specs/rules (to match Squid code). Due to reasons detailed in my previous message, we cannot simply do nothing (i.e. let document producers decide whether to produce such mixed reports).

Since Squid was designed to support mixed reports, your tests show that they more-or-less work, and they may be useful, I think we should adjust the specs. I will post an adjusted summary in a dedicated comment.

Reply:
Here is the updated proposal. It simplifies stream syntax requirements and allows mixed reports. It does not require any serious code rewrites AFAICT. Did I miss any important bits?

----

When transition to YAML is over, each successful cache manager response body (a.k.a. "report") will have `l-yaml-stream` syntax that supports zero or more documents per report. Most (and possibly all) reports will use the following subset of that rich syntax, using `---` and `...` document boundary markers:

```bnf
cache-manager-report-typical ::= ( l-explicit-document l-document-suffix )*
```

Within a report:

1: A kid-specific document produced by a kid process is a YAML map containing a `kid: ID` entry (specifying the ID of the kid process that produced the document) and no `kids: ...` entry.

2: A document produced by Coordinator by aggregating SMP kid information is a YAML map containing a `kids: [ ... ]` entry (specifying zero or more IDs of kid processes that contributed information to the aggregated document) and no `kid: ID` entry.

3: Any other other document (e.g., a single document returned in response to a "mgr:menu" request) will either be a YAML map with no `kid` entry and no `kids` entry OR it will not be a YAML map.

Reply:
Almost works for me, except:

- why focusing on YAML maps? It's a good abstraction and it works well with the "kids" framing, but it might have to be an array or another schema. So wherever you say "yaml map" I would specify "valid yaml" instead
- I would rephrase point 3 to:
> 3: Any other other document (e.g., a single document returned in response to a "mgr:menu" request) will either be a YAML map with no kid entry and no kids entry OR it will not be considered YAML at all

(alternatives such as "bound to be YAML") are equally good)

Reply:
> - why focusing on YAML maps?

... because a consumer cannot extract kid(s) entries from any other structure. The underlying assumption here is that consumers ought to be able to process these reports, at least at some basic document classification and rendering level, without hard-coded knowledge of each report meaning.

> It's a good abstraction and it works well with the "kids" framing,

That is why it is there for documents that require such framing.

> it might have to be an array or another schema. So wherever you say "yaml map" I would specify "valid yaml" instead

That is already done for all "other" documents that do not need kid(s) entries.

> Rephrased 3: Any other other document (e.g., a single document returned in response to a "mgr:menu" request) will either be a YAML map with no kid entry and no kids entry OR it will not be considered YAML at all

If it is not YAML, it is not subject to these rules. The YAML vs not-YAML decision is done using Content-Type, _before_ these rules are applied.

Also please note the "all reports have been converted to YAML" precondition. We can drop that, of course, but the rules will still only apply to YAML reports.

Do these clarifications address your concerns? Any other text changes we should make?

Reply:
Yes, both. I am closing this thread and posting the corresponding review.